// @teambit/react-native/dist/templates/react-native-env/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.reactNativeTemplate = void 0;

function _index() {
  const data = require("./files/index");

  _index = function () {
    return data;
  };

  return data;
}

function _doc() {
  const data = require("./files/doc");

  _doc = function () {
    return data;
  };

  return data;
}

function _extension() {
  const data = require("./files/extension");

  _extension = function () {
    return data;
  };

  return data;
}

const reactNativeTemplate = {
  name: 'react-native-env',
  description: 'customize the base React Native env with your configs and tools',
  generateFiles: context => {
    return [{
      relativePath: 'index.ts',
      content: (0, _index().indexFile)(context),
      isMain: true
    }, {
      relativePath: `${context.name}.docs.mdx`,
      content: (0, _doc().docFile)()
    }, {
      relativePath: `${context.name}.extension.ts`,
      content: (0, _extension().extensionFile)(context)
    }];
  }
};
exports.reactNativeTemplate = reactNativeTemplate;
