// @teambit/react/dist/templates/react-context/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.reactContext = void 0;

function _context() {
  const data = require("./files/context");

  _context = function () {
    return data;
  };

  return data;
}

function _contextProvider() {
  const data = require("./files/context-provider");

  _contextProvider = function () {
    return data;
  };

  return data;
}

function _composition() {
  const data = require("./files/composition");

  _composition = function () {
    return data;
  };

  return data;
}

function _docs() {
  const data = require("./files/docs");

  _docs = function () {
    return data;
  };

  return data;
}

function _test() {
  const data = require("./files/test");

  _test = function () {
    return data;
  };

  return data;
}

const reactContext = {
  name: 'react-context',
  description: 'a react context component',
  generateFiles: context => {
    const {
      name,
      namePascalCase: Name
    } = context;
    const indexFile = {
      relativePath: 'index.ts',
      content: `export { ${Name}Context } from './${name}-context';
export type { ${Name}ContextType } from './${name}-context';
export { ${Name}Provider } from './${name}-context-provider';
export type { ${Name}ProviderProps } from './${name}-context-provider';
`
    };
    return [indexFile, (0, _context().contextFile)(context), (0, _contextProvider().contextProviderFile)(context), (0, _composition().compositionFile)(context), (0, _docs().docsFile)(context), (0, _test().testFile)(context)];
  }
};
exports.reactContext = reactContext;
