// @teambit/react-router/dist/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "ReactRouterAspect", {
  enumerable: true,
  get: function () {
    return _reactRouter().ReactRouterAspect;
  }
});
Object.defineProperty(exports, "Routing", {
  enumerable: true,
  get: function () {
    return _routingMethod().Routing;
  }
});
Object.defineProperty(exports, "Link", {
  enumerable: true,
  get: function () {
    return _baseUiRouting().Link;
  }
});
Object.defineProperty(exports, "LinkProps", {
  enumerable: true,
  get: function () {
    return _baseUiRouting().LinkProps;
  }
});
Object.defineProperty(exports, "NavLink", {
  enumerable: true,
  get: function () {
    return _baseUiRouting2().NavLink;
  }
});
Object.defineProperty(exports, "NavLinkProps", {
  enumerable: true,
  get: function () {
    return _baseUiRouting2().NavLinkProps;
  }
});
Object.defineProperty(exports, "LinkAnchor", {
  enumerable: true,
  get: function () {
    return _uiFoundationUiReactRouter().LinkAnchor;
  }
});
Object.defineProperty(exports, "LinkContextProvider", {
  enumerable: true,
  get: function () {
    return _uiFoundationUiReactRouter().LinkContextProvider;
  }
});
Object.defineProperty(exports, "useLinkContext", {
  enumerable: true,
  get: function () {
    return _uiFoundationUiReactRouter().useLinkContext;
  }
});
exports.default = void 0;

function _reactRouter() {
  const data = require("./react-router.aspect");

  _reactRouter = function () {
    return data;
  };

  return data;
}

function _routingMethod() {
  const data = require("./routing-method");

  _routingMethod = function () {
    return data;
  };

  return data;
}

function _baseUiRouting() {
  const data = require("@teambit/base-ui.routing.link");

  _baseUiRouting = function () {
    return data;
  };

  return data;
}

function _baseUiRouting2() {
  const data = require("@teambit/base-ui.routing.nav-link");

  _baseUiRouting2 = function () {
    return data;
  };

  return data;
}

function _uiFoundationUiReactRouter() {
  const data = require("@teambit/ui-foundation.ui.react-router.link-anchor");

  _uiFoundationUiReactRouter = function () {
    return data;
  };

  return data;
}

// TODO - replace with resolveFromHost / resolveFromApp when available
// this exposes the Link components installed in bit bin, so they can use the same RoutingProvider file from node_modules
var _default = _reactRouter().ReactRouterAspect;

exports.default = _default;
