// @teambit/config/dist/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Config", {
  enumerable: true,
  get: function () {
    return _config().Config;
  }
});
Object.defineProperty(exports, "ConfigAspect", {
  enumerable: true,
  get: function () {
    return _config2().ConfigAspect;
  }
});
Object.defineProperty(exports, "ConfigRuntime", {
  enumerable: true,
  get: function () {
    return _config2().ConfigRuntime;
  }
});
Object.defineProperty(exports, "ComponentScopeDirMap", {
  enumerable: true,
  get: function () {
    return _workspaceConfig().ComponentScopeDirMap;
  }
});
Object.defineProperty(exports, "getWorkspaceConfigTemplateParsed", {
  enumerable: true,
  get: function () {
    return _workspaceConfig().getWorkspaceConfigTemplateParsed;
  }
});
Object.defineProperty(exports, "stringifyWorkspaceConfig", {
  enumerable: true,
  get: function () {
    return _workspaceConfig().stringifyWorkspaceConfig;
  }
});

function _config() {
  const data = require("./config");

  _config = function () {
    return data;
  };

  return data;
}

function _config2() {
  const data = require("./config.aspect");

  _config2 = function () {
    return data;
  };

  return data;
}

function _workspaceConfig() {
  const data = require("./workspace-config");

  _workspaceConfig = function () {
    return data;
  };

  return data;
}
