// @teambit/dependency-resolver/dist/manifest/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "ManifestDependenciesKeys", {
  enumerable: true,
  get: function () {
    return _manifest().ManifestDependenciesKeys;
  }
});
Object.defineProperty(exports, "ManifestDependenciesObject", {
  enumerable: true,
  get: function () {
    return _manifest().ManifestDependenciesObject;
  }
});
Object.defineProperty(exports, "ComponentManifest", {
  enumerable: true,
  get: function () {
    return _componentManifest().ComponentManifest;
  }
});
Object.defineProperty(exports, "WorkspaceManifest", {
  enumerable: true,
  get: function () {
    return _workspaceManifest().WorkspaceManifest;
  }
});
Object.defineProperty(exports, "WorkspaceManifestFactory", {
  enumerable: true,
  get: function () {
    return _workspaceManifestFactory().WorkspaceManifestFactory;
  }
});
Object.defineProperty(exports, "CreateFromComponentsOptions", {
  enumerable: true,
  get: function () {
    return _workspaceManifestFactory().CreateFromComponentsOptions;
  }
});
Object.defineProperty(exports, "DepsFilterFn", {
  enumerable: true,
  get: function () {
    return _workspaceManifestFactory().DepsFilterFn;
  }
});

function _manifest() {
  const data = require("./manifest");

  _manifest = function () {
    return data;
  };

  return data;
}

function _componentManifest() {
  const data = require("./component-manifest");

  _componentManifest = function () {
    return data;
  };

  return data;
}

function _workspaceManifest() {
  const data = require("./workspace-manifest");

  _workspaceManifest = function () {
    return data;
  };

  return data;
}

function _workspaceManifestFactory() {
  const data = require("./workspace-manifest-factory");

  _workspaceManifestFactory = function () {
    return data;
  };

  return data;
}
