// @teambit/bundler/dist/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  DevServer: true,
  BundlerContext: true,
  Target: true,
  DevServerContext: true,
  Bundler: true,
  BundlerResult: true,
  BundlerMode: true,
  BundlerAspect: true,
  ComponentDir: true,
  ComponentServer: true
};
Object.defineProperty(exports, "DevServer", {
  enumerable: true,
  get: function () {
    return _devServer().DevServer;
  }
});
Object.defineProperty(exports, "BundlerContext", {
  enumerable: true,
  get: function () {
    return _devServerContext().BundlerContext;
  }
});
Object.defineProperty(exports, "Target", {
  enumerable: true,
  get: function () {
    return _devServerContext().Target;
  }
});
Object.defineProperty(exports, "DevServerContext", {
  enumerable: true,
  get: function () {
    return _devServerContext().DevServerContext;
  }
});
Object.defineProperty(exports, "Bundler", {
  enumerable: true,
  get: function () {
    return _bundler().Bundler;
  }
});
Object.defineProperty(exports, "BundlerResult", {
  enumerable: true,
  get: function () {
    return _bundler().BundlerResult;
  }
});
Object.defineProperty(exports, "BundlerMode", {
  enumerable: true,
  get: function () {
    return _bundler().BundlerMode;
  }
});
Object.defineProperty(exports, "BundlerAspect", {
  enumerable: true,
  get: function () {
    return _bundler2().BundlerAspect;
  }
});
Object.defineProperty(exports, "ComponentDir", {
  enumerable: true,
  get: function () {
    return _getEntry().ComponentDir;
  }
});
Object.defineProperty(exports, "ComponentServer", {
  enumerable: true,
  get: function () {
    return _componentServer().ComponentServer;
  }
});

function _devServer() {
  const data = require("./dev-server");

  _devServer = function () {
    return data;
  };

  return data;
}

function _devServerContext() {
  const data = require("./dev-server-context");

  _devServerContext = function () {
    return data;
  };

  return data;
}

function _bundler() {
  const data = require("./bundler");

  _bundler = function () {
    return data;
  };

  return data;
}

function _bundler2() {
  const data = require("./bundler.aspect");

  _bundler2 = function () {
    return data;
  };

  return data;
}

function _getEntry() {
  const data = require("./get-entry");

  _getEntry = function () {
    return data;
  };

  return data;
}

function _componentServer() {
  const data = require("./component-server");

  _componentServer = function () {
    return data;
  };

  return data;
}

var _events = require("./events");

Object.keys(_events).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _events[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _events[key];
    }
  });
});
