// @teambit/docs/dist/doc/doc.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Doc = void 0;

function _docPropList() {
  const data = require("./doc-prop-list");

  _docPropList = function () {
    return data;
  };

  return data;
}

class Doc {
  constructor(filePath, props) {
    this.filePath = filePath;
    this.props = props;
  }

  toObject() {
    return {
      filePath: this.filePath,
      props: this.props.docProps
    };
  }
  /**
   * shorthand for getting the component description.
   */


  get description() {
    var _this$props$get;

    const value = (_this$props$get = this.props.get('description')) === null || _this$props$get === void 0 ? void 0 : _this$props$get.value;
    if (!value) return '';
    return value;
  }
  /**
   * shorthand for getting the component labels.
   */


  get labels() {
    var _this$props$get2;

    const value = (_this$props$get2 = this.props.get('labels')) === null || _this$props$get2 === void 0 ? void 0 : _this$props$get2.value;
    if (!value) return [];
    return value;
  }
  /**
   * shorthand for getting the component labels.
   */


  get displayName() {
    return this.props.get('displayName');
  }

  static from(path, propObject) {
    return new Doc(path, _docPropList().DocPropList.from(propObject));
  }

}

exports.Doc = Doc;
