// @teambit/aspect-loader/dist/core-aspects.js
"use strict";

require("core-js/modules/es.array.iterator.js");

require("core-js/modules/es.promise.js");

require("core-js/modules/es.regexp.exec.js");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getAspectDir = getAspectDir;
exports.getAspectDistDir = getAspectDistDir;
exports.getCoreAspectName = getCoreAspectName;
exports.getCoreAspectPackageName = getCoreAspectPackageName;
exports.getAspectDef = getAspectDef;

function _fsExtra() {
  const data = require("fs-extra");

  _fsExtra = function () {
    return data;
  };

  return data;
}

function _path() {
  const data = require("path");

  _path = function () {
    return data;
  };

  return data;
}

function getAspectDir(id) {
  const aspectName = getCoreAspectName(id);
  const packageName = getCoreAspectPackageName(id);
  let dirPath;

  try {
    const moduleDirectory = require.resolve(packageName);

    dirPath = (0, _path().join)(moduleDirectory, '../..'); // to remove the "index.js" at the end
  } catch (err) {
    dirPath = (0, _path().resolve)(__dirname, '../..', aspectName, 'dist');
  }

  if (!(0, _fsExtra().existsSync)(dirPath)) {
    throw new Error(`unable to find ${aspectName} in ${dirPath}`);
  }

  return dirPath;
}

function getAspectDistDir(id) {
  return (0, _path().resolve)(`${getAspectDir(id)}/dist`);
}

function getCoreAspectName(id) {
  const [, ...name] = id.split('/');
  const aspectName = name.join('.');
  return aspectName;
}

function getCoreAspectPackageName(id) {
  const aspectName = getCoreAspectName(id);
  return `@teambit/${aspectName}`;
}

async function getAspectDef(aspectName, runtime) {
  const dirPath = getAspectDistDir(aspectName);
  const files = await (0, _fsExtra().readdir)(dirPath);
  let runtimeFile;

  if (runtime) {
    runtimeFile = files.find(file => file.includes(`.${runtime}.runtime.js`)) || null;
  }

  return {
    id: aspectName,
    aspectPath: (0, _path().join)(dirPath, '..'),
    runtimePath: runtimeFile ? (0, _path().resolve)(`${dirPath}/${runtimeFile}`) : null
  };
}
