// @teambit/aspect-loader/dist/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "AspectLoaderAspect", {
  enumerable: true,
  get: function () {
    return _aspectLoader().AspectLoaderAspect;
  }
});
Object.defineProperty(exports, "getAspectDef", {
  enumerable: true,
  get: function () {
    return _coreAspects().getAspectDef;
  }
});
Object.defineProperty(exports, "getAspectDir", {
  enumerable: true,
  get: function () {
    return _coreAspects().getAspectDir;
  }
});
Object.defineProperty(exports, "getAspectDistDir", {
  enumerable: true,
  get: function () {
    return _coreAspects().getAspectDistDir;
  }
});
Object.defineProperty(exports, "getCoreAspectPackageName", {
  enumerable: true,
  get: function () {
    return _coreAspects().getCoreAspectPackageName;
  }
});
Object.defineProperty(exports, "getCoreAspectName", {
  enumerable: true,
  get: function () {
    return _coreAspects().getCoreAspectName;
  }
});
Object.defineProperty(exports, "AspectDefinition", {
  enumerable: true,
  get: function () {
    return _aspectDefinition().AspectDefinition;
  }
});
exports.default = void 0;

function _aspectLoader() {
  const data = require("./aspect-loader.aspect");

  _aspectLoader = function () {
    return data;
  };

  return data;
}

function _coreAspects() {
  const data = require("./core-aspects");

  _coreAspects = function () {
    return data;
  };

  return data;
}

function _aspectDefinition() {
  const data = require("./aspect-definition");

  _aspectDefinition = function () {
    return data;
  };

  return data;
}

var _default = _aspectLoader().AspectLoaderAspect;

exports.default = _default;
