// @teambit/ui/dist/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  UIAspect: true,
  UIRuntime: true,
  UIRoot: true,
  PostStartOptions: true,
  ProxyEntry: true,
  StartPlugin: true,
  StartPluginOptions: true,
  UIRootUI: true,
  UIRootFactory: true,
  DataQueryResult: true,
  useDataQuery: true,
  useMutation: true
};
Object.defineProperty(exports, "UIAspect", {
  enumerable: true,
  get: function () {
    return _ui().UIAspect;
  }
});
Object.defineProperty(exports, "UIRuntime", {
  enumerable: true,
  get: function () {
    return _ui().UIRuntime;
  }
});
Object.defineProperty(exports, "UIRoot", {
  enumerable: true,
  get: function () {
    return _uiRoot().UIRoot;
  }
});
Object.defineProperty(exports, "PostStartOptions", {
  enumerable: true,
  get: function () {
    return _uiRoot().PostStartOptions;
  }
});
Object.defineProperty(exports, "ProxyEntry", {
  enumerable: true,
  get: function () {
    return _uiRoot().ProxyEntry;
  }
});
Object.defineProperty(exports, "StartPlugin", {
  enumerable: true,
  get: function () {
    return _startPlugin().StartPlugin;
  }
});
Object.defineProperty(exports, "StartPluginOptions", {
  enumerable: true,
  get: function () {
    return _startPlugin().StartPluginOptions;
  }
});
Object.defineProperty(exports, "UIRootUI", {
  enumerable: true,
  get: function () {
    return _uiRoot2().UIRootUI;
  }
});
Object.defineProperty(exports, "UIRootFactory", {
  enumerable: true,
  get: function () {
    return _uiRoot2().UIRootFactory;
  }
});
Object.defineProperty(exports, "DataQueryResult", {
  enumerable: true,
  get: function () {
    return _uiFoundationUiHooks().DataQueryResult;
  }
});
Object.defineProperty(exports, "useDataQuery", {
  enumerable: true,
  get: function () {
    return _uiFoundationUiHooks().useDataQuery;
  }
});
Object.defineProperty(exports, "useMutation", {
  enumerable: true,
  get: function () {
    return _client().useMutation;
  }
});
exports.default = void 0;

function _ui() {
  const data = require("./ui.aspect");

  _ui = function () {
    return data;
  };

  return data;
}

var _events = require("./events");

Object.keys(_events).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _events[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _events[key];
    }
  });
});

function _uiRoot() {
  const data = require("./ui-root");

  _uiRoot = function () {
    return data;
  };

  return data;
}

function _startPlugin() {
  const data = require("./start-plugin");

  _startPlugin = function () {
    return data;
  };

  return data;
}

function _uiRoot2() {
  const data = require("./ui-root.ui");

  _uiRoot2 = function () {
    return data;
  };

  return data;
}

function _uiFoundationUiHooks() {
  const data = require("@teambit/ui-foundation.ui.hooks.use-data-query");

  _uiFoundationUiHooks = function () {
    return data;
  };

  return data;
}

function _client() {
  const data = require("@apollo/client");

  _client = function () {
    return data;
  };

  return data;
}

var _default = _ui().UIAspect; // using `useDataQuery` from this package is deprecated, use `@teambit/ui-foundation.ui.hooks.use-data-query` directly


exports.default = _default;
