// @teambit/component/dist/ui/menu/menu.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

require("core-js/modules/es.array.iterator.js");

require("core-js/modules/es.regexp.exec.js");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Menu = Menu;

function _uiFoundationUi() {
  const data = require("@teambit/ui-foundation.ui.main-dropdown");

  _uiFoundationUi = function () {
    return data;
  };

  return data;
}

function _componentUi() {
  const data = require("@teambit/component.ui.version-dropdown");

  _componentUi = function () {
    return data;
  };

  return data;
}

function _fullLoader() {
  const data = require("@teambit/legacy/dist/to-eject/full-loader");

  _fullLoader = function () {
    return data;
  };

  return data;
}

function _lodash() {
  const data = require("lodash");

  _lodash = function () {
    return data;
  };

  return data;
}

function _classnames() {
  const data = _interopRequireDefault(require("classnames"));

  _classnames = function () {
    return data;
  };

  return data;
}

function _react() {
  const data = _interopRequireWildcard(require("react"));

  _react = function () {
    return data;
  };

  return data;
}

function _uiFoundationUiUseBox() {
  const data = require("@teambit/ui-foundation.ui.use-box.dropdown");

  _uiFoundationUiUseBox = function () {
    return data;
  };

  return data;
}

function _uiFoundationUiUseBox2() {
  const data = require("@teambit/ui-foundation.ui.use-box.menu");

  _uiFoundationUiUseBox2 = function () {
    return data;
  };

  return data;
}

function _useComponent() {
  const data = require("../use-component");

  _useComponent = function () {
    return data;
  };

  return data;
}

function _menuNav() {
  const data = require("./menu-nav");

  _menuNav = function () {
    return data;
  };

  return data;
}

function _menuModule() {
  const data = _interopRequireDefault(require("./menu.module.scss"));

  _menuModule = function () {
    return data;
  };

  return data;
}

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

/**
 * top bar menu.
 */
function Menu({
  navigationSlot,
  widgetSlot,
  className,
  host,
  menuItemSlot
}) {
  const {
    component
  } = (0, _useComponent().useComponent)(host);
  const mainMenuItems = (0, _react().useMemo)(() => (0, _lodash().groupBy)((0, _lodash().flatten)(menuItemSlot.values()), 'category'), [menuItemSlot]);
  if (!component) return /*#__PURE__*/_react().default.createElement(_fullLoader().FullLoader, null);
  return /*#__PURE__*/_react().default.createElement("div", {
    className: (0, _classnames().default)(_menuModule().default.topBar, className)
  }, /*#__PURE__*/_react().default.createElement("div", {
    className: _menuModule().default.leftSide
  }, /*#__PURE__*/_react().default.createElement(_menuNav().MenuNav, {
    navigationSlot: navigationSlot
  })), /*#__PURE__*/_react().default.createElement("div", {
    className: _menuModule().default.rightSide
  }, /*#__PURE__*/_react().default.createElement("div", {
    className: _menuModule().default.widgets
  }, /*#__PURE__*/_react().default.createElement(_menuNav().MenuNav, {
    navigationSlot: widgetSlot
  })), /*#__PURE__*/_react().default.createElement(VersionRelatedDropdowns, {
    component: component
  }), /*#__PURE__*/_react().default.createElement(_uiFoundationUi().MainDropdown, {
    menuItems: mainMenuItems
  })));
}

function VersionRelatedDropdowns({
  component
}) {
  const versionList = (0, _react().useMemo)(() => {
    var _component$tags;

    return (_component$tags = component.tags) === null || _component$tags === void 0 ? void 0 : _component$tags.toArray().map(tag => {
      var _tag$version;

      return tag === null || tag === void 0 ? void 0 : (_tag$version = tag.version) === null || _tag$version === void 0 ? void 0 : _tag$version.version;
    }).filter(x => x !== undefined).reverse();
  }, [component.tags]) || [];
  const isLatestVersion = (0, _react().useMemo)(() => component.version === versionList[0], [component.version]);
  const packageVersion = (0, _react().useMemo)(() => isLatestVersion ? '' : `@${component.version}`, [component.version]);
  return /*#__PURE__*/_react().default.createElement(_react().default.Fragment, null, versionList.length > 0 && /*#__PURE__*/_react().default.createElement(_uiFoundationUiUseBox().UseBoxDropdown, {
    position: "bottom-end",
    className: _menuModule().default.useBox,
    Menu: () => /*#__PURE__*/_react().default.createElement(_uiFoundationUiUseBox2().Menu, {
      componentName: component.id.name,
      componentId: component.id.toString({
        ignoreVersion: isLatestVersion
      }),
      packageName: `${component.packageName}${packageVersion}`,
      registryName: component.packageName.split('/')[0]
    })
  }), /*#__PURE__*/_react().default.createElement(_componentUi().VersionDropdown, {
    versions: versionList,
    currentVersion: component.version
  }));
}
