// @teambit/preview/dist/strategies/component-strategy.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

require("core-js/modules/es.array.iterator.js");

require("core-js/modules/es.promise.js");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ComponentBundlingStrategy = void 0;

function _defineProperty2() {
  const data = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

  _defineProperty2 = function () {
    return data;
  };

  return data;
}

class ComponentBundlingStrategy {
  constructor() {
    (0, _defineProperty2().default)(this, "name", 'component');
  }

  computeTargets(context, previewDefs, previewTask) {
    return Promise.all(context.capsuleNetwork.graphCapsules.map(async capsule => {
      return {
        entries: await previewTask.computePaths(capsule, previewDefs, context),
        components: [capsule.component],
        outputPath: capsule.path
      };
    }));
  }

  async computeResults(context, results, previewTask) {
    return {
      componentsResults: results.map(result => {
        return {
          errors: result.errors,
          component: result.components[0],
          warning: result.warnings
        };
      }),
      artifacts: [{
        name: 'preview',
        globPatterns: [previewTask.getPreviewDirectory(context)]
      }]
    };
  }

}

exports.ComponentBundlingStrategy = ComponentBundlingStrategy;
