// @teambit/builder/dist/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "BuildPipe", {
  enumerable: true,
  get: function () {
    return _buildPipe().BuildPipe;
  }
});
Object.defineProperty(exports, "ComponentResult", {
  enumerable: true,
  get: function () {
    return _types().ComponentResult;
  }
});
Object.defineProperty(exports, "TaskMetadata", {
  enumerable: true,
  get: function () {
    return _types().TaskMetadata;
  }
});
Object.defineProperty(exports, "BuildContext", {
  enumerable: true,
  get: function () {
    return _buildTask().BuildContext;
  }
});
Object.defineProperty(exports, "BuildTask", {
  enumerable: true,
  get: function () {
    return _buildTask().BuildTask;
  }
});
Object.defineProperty(exports, "BuiltTaskResult", {
  enumerable: true,
  get: function () {
    return _buildTask().BuiltTaskResult;
  }
});
Object.defineProperty(exports, "TaskLocation", {
  enumerable: true,
  get: function () {
    return _buildTask().TaskLocation;
  }
});
Object.defineProperty(exports, "BuildTaskHelper", {
  enumerable: true,
  get: function () {
    return _buildTask().BuildTaskHelper;
  }
});
Object.defineProperty(exports, "BuilderAspect", {
  enumerable: true,
  get: function () {
    return _builder().BuilderAspect;
  }
});
Object.defineProperty(exports, "StorageResolver", {
  enumerable: true,
  get: function () {
    return _storage().StorageResolver;
  }
});
Object.defineProperty(exports, "Artifact", {
  enumerable: true,
  get: function () {
    return _artifact().Artifact;
  }
});
Object.defineProperty(exports, "ArtifactList", {
  enumerable: true,
  get: function () {
    return _artifact().ArtifactList;
  }
});
Object.defineProperty(exports, "ArtifactFactory", {
  enumerable: true,
  get: function () {
    return _artifact().ArtifactFactory;
  }
});
Object.defineProperty(exports, "ArtifactDefinition", {
  enumerable: true,
  get: function () {
    return _artifact().ArtifactDefinition;
  }
});
Object.defineProperty(exports, "TaskResultsList", {
  enumerable: true,
  get: function () {
    return _taskResultsList().TaskResultsList;
  }
});

function _buildPipe() {
  const data = require("./build-pipe");

  _buildPipe = function () {
    return data;
  };

  return data;
}

function _types() {
  const data = require("./types");

  _types = function () {
    return data;
  };

  return data;
}

function _buildTask() {
  const data = require("./build-task");

  _buildTask = function () {
    return data;
  };

  return data;
}

function _builder() {
  const data = require("./builder.aspect");

  _builder = function () {
    return data;
  };

  return data;
}

function _storage() {
  const data = require("./storage");

  _storage = function () {
    return data;
  };

  return data;
}

function _artifact() {
  const data = require("./artifact");

  _artifact = function () {
    return data;
  };

  return data;
}

function _taskResultsList() {
  const data = require("./task-results-list");

  _taskResultsList = function () {
    return data;
  };

  return data;
}
