// @teambit/deprecation/dist/deprecation.fragment.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

require("core-js/modules/es.promise.js");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DeprecationFragment = void 0;

function _defineProperty2() {
  const data = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

  _defineProperty2 = function () {
    return data;
  };

  return data;
}

class DeprecationFragment {
  constructor(deprecation) {
    this.deprecation = deprecation;
    (0, _defineProperty2().default)(this, "title", 'deprecated');
    (0, _defineProperty2().default)(this, "weight", 3);
  }

  async renderRow(component) {
    return {
      title: this.title,
      content: this.deprecation.getDeprecationInfo(component).isDeprecate.toString()
    };
  }

  async json(component) {
    return {
      title: this.title,
      json: this.deprecation.getDeprecationInfo(component)
    };
  }

}

exports.DeprecationFragment = DeprecationFragment;
