// @teambit/docs/dist/overview/overview.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Overview = Overview;

function _react() {
  const data = _interopRequireWildcard(require("react"));

  _react = function () {
    return data;
  };

  return data;
}

function _component() {
  const data = require("@teambit/component");

  _component = function () {
    return data;
  };

  return data;
}

function _previewUi() {
  const data = require("@teambit/preview.ui.component-preview");

  _previewUi = function () {
    return data;
  };

  return data;
}

function _designUiSurfaces() {
  const data = require("@teambit/design.ui.surfaces.status-message-card");

  _designUiSurfaces = function () {
    return data;
  };

  return data;
}

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function Overview() {
  const component = (0, _react().useContext)(_component().ComponentContext);
  if ((component === null || component === void 0 ? void 0 : component.buildStatus) === 'pending' && (component === null || component === void 0 ? void 0 : component.host) === 'teambit.scope/scope') return /*#__PURE__*/_react().default.createElement(_designUiSurfaces().StatusMessageCard, {
    style: {
      margin: 'auto'
    },
    status: "PROCESSING",
    title: "component preview pending"
  }, "this might take some time");
  if ((component === null || component === void 0 ? void 0 : component.buildStatus) === 'failed' && (component === null || component === void 0 ? void 0 : component.host) === 'teambit.scope/scope') return /*#__PURE__*/_react().default.createElement(_designUiSurfaces().StatusMessageCard, {
    style: {
      margin: 'auto'
    },
    status: "FAILURE",
    title: "failed to get component preview "
  });
  return /*#__PURE__*/_react().default.createElement(_previewUi().ComponentPreview, {
    component: component,
    style: {
      width: '100%',
      height: '100%'
    },
    previewName: "overview"
  });
}
