// @teambit/react/dist/templates/react-hook/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.reactHook = void 0;

function _component() {
  const data = require("./files/component");

  _component = function () {
    return data;
  };

  return data;
}

function _composition() {
  const data = require("./files/composition");

  _composition = function () {
    return data;
  };

  return data;
}

function _docs() {
  const data = require("./files/docs");

  _docs = function () {
    return data;
  };

  return data;
}

function _test() {
  const data = require("./files/test");

  _test = function () {
    return data;
  };

  return data;
}

const reactHook = {
  name: 'react-hook',
  description: 'a react hook component',
  generateFiles: context => {
    const {
      name,
      nameCamelCase: Name
    } = context;
    const indexFile = {
      relativePath: 'index.ts',
      content: `export { ${Name} } from './${name}';
`
    };
    return [indexFile, (0, _component().componentFile)(context), (0, _composition().compositionFile)(context), (0, _docs().docsFile)(context), (0, _test().testFile)(context)];
  }
};
exports.reactHook = reactHook;
