// @teambit/sidebar/dist/ui/side-bar/side-bar.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

require("core-js/modules/es.array.iterator.js");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SideBar = SideBar;

function _extends2() {
  const data = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

  _extends2 = function () {
    return data;
  };

  return data;
}

function _objectWithoutProperties2() {
  const data = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

  _objectWithoutProperties2 = function () {
    return data;
  };

  return data;
}

function _react() {
  const data = _interopRequireWildcard(require("react"));

  _react = function () {
    return data;
  };

  return data;
}

function _lodash() {
  const data = require("lodash");

  _lodash = function () {
    return data;
  };

  return data;
}

function _designUiSurfacesMenu() {
  const data = require("@teambit/design.ui.surfaces.menu.section");

  _designUiSurfacesMenu = function () {
    return data;
  };

  return data;
}

function _uiFoundationUiTree() {
  const data = require("@teambit/ui-foundation.ui.tree.drawer");

  _uiFoundationUiTree = function () {
    return data;
  };

  return data;
}

function _sideBarModule() {
  const data = _interopRequireDefault(require("./side-bar.module.scss"));

  _sideBarModule = function () {
    return data;
  };

  return data;
}

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

/**
 * side bar component.
 */
function SideBar(_ref) {
  let {
    drawerSlot,
    itemSlot
  } = _ref,
      rest = (0, _objectWithoutProperties2().default)(_ref, ["drawerSlot", "itemSlot"]);
  const [openDrawerList, onToggleDrawer] = (0, _react().useState)([drawerSlot.toArray()[0][0]]);
  const items = (0, _react().useMemo)(() => (0, _lodash().flatten)(itemSlot === null || itemSlot === void 0 ? void 0 : itemSlot.values()), [itemSlot]);

  const handleDrawerToggle = id => {
    const isDrawerOpen = openDrawerList.includes(id);

    if (isDrawerOpen) {
      onToggleDrawer(list => list.filter(drawer => drawer !== id));
      return;
    }

    onToggleDrawer(list => list.concat(id));
  };

  return /*#__PURE__*/_react().default.createElement("div", (0, _extends2().default)({}, rest, {
    className: _sideBarModule().default.sidebar
  }), /*#__PURE__*/_react().default.createElement(_designUiSurfacesMenu().MenuSection, {
    items: items
  }), drawerSlot.toArray().map(([id, drawer]) => {
    if (!drawer || !drawer.name) return null;
    return /*#__PURE__*/_react().default.createElement(_uiFoundationUiTree().DrawerUI, {
      isOpen: openDrawerList.includes(id),
      onToggle: () => handleDrawerToggle(id),
      key: id,
      name: drawer.name
    }, /*#__PURE__*/_react().default.createElement(drawer.render, null));
  }));
}
