// @teambit/react/dist/templates/react-component/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.deprecatedReactComponent = exports.reactComponent = void 0;

function _component() {
  const data = require("./files/component");

  _component = function () {
    return data;
  };

  return data;
}

function _composition() {
  const data = require("./files/composition");

  _composition = function () {
    return data;
  };

  return data;
}

function _docs() {
  const data = require("./files/docs");

  _docs = function () {
    return data;
  };

  return data;
}

function _test() {
  const data = require("./files/test");

  _test = function () {
    return data;
  };

  return data;
}

const reactComponent = {
  name: 'react',
  description: 'a basic react component',
  generateFiles: context => {
    const {
      name,
      namePascalCase: Name
    } = context;
    const indexFile = {
      relativePath: 'index.ts',
      content: `export { ${Name} } from './${name}';
export type { ${Name}Props } from './${name}';
`
    };
    return [indexFile, (0, _component().componentFile)(context), (0, _composition().compositionFile)(context), (0, _docs().docsFile)(context), (0, _test().testFile)(context)];
  }
};
exports.reactComponent = reactComponent;
const deprecatedReactComponent = {
  name: 'react-component',
  description: 'a basic react component',
  hidden: true,
  generateFiles: context => {
    const {
      name,
      namePascalCase: Name
    } = context;
    const indexFile = {
      relativePath: 'index.ts',
      content: `export { ${Name} } from './${name}';
export type { ${Name}Props } from './${name}';
`
    };
    return [indexFile, (0, _component().componentFile)(context), (0, _composition().compositionFile)(context), (0, _docs().docsFile)(context), (0, _test().testFile)(context)];
  }
};
exports.deprecatedReactComponent = deprecatedReactComponent;
