// @teambit/workspace/dist/ui/workspace/workspace-model.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Workspace = void 0;

function _component() {
  const data = require("@teambit/component");

  _component = function () {
    return data;
  };

  return data;
}

// import { ComponentMeta } from '@teambit/component';
class Workspace {
  constructor(
  /**
   * name of the workspace.
   */
  name,
  /**
   * absolute path of the workspace.
   */
  path,
  /**
   * icon of the workspace.
   */
  icon,
  /**
   * components container in the workspace.
   */
  components) {
    this.name = name;
    this.path = path;
    this.icon = icon;
    this.components = components;
  }
  /**
   * return a component from the workspace.
   */


  getComponent(id) {
    return this.components.find(component => component.id.fullName === id.fullName);
  }

  static from({
    name,
    path,
    components,
    icon
  }) {
    return new Workspace(name, path, icon, components.map(value => {
      return _component().ComponentModel.from(value);
    }));
  }

  static empty() {
    return new Workspace('', '', '', []);
  }

}

exports.Workspace = Workspace;
