// @teambit/scope/dist/ui/components-drawer/components.drawer.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ComponentsDrawer = void 0;

function _extends2() {
  const data = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

  _extends2 = function () {
    return data;
  };

  return data;
}

function _defineProperty2() {
  const data = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

  _defineProperty2 = function () {
    return data;
  };

  return data;
}

function _react() {
  const data = _interopRequireWildcard(require("react"));

  _react = function () {
    return data;
  };

  return data;
}

function _classnames() {
  const data = _interopRequireDefault(require("classnames"));

  _classnames = function () {
    return data;
  };

  return data;
}

function _uiFoundationUi() {
  const data = require("@teambit/ui-foundation.ui.side-bar");

  _uiFoundationUi = function () {
    return data;
  };

  return data;
}

function _uiFoundationUi2() {
  const data = require("@teambit/ui-foundation.ui.full-loader");

  _uiFoundationUi2 = function () {
    return data;
  };

  return data;
}

function _designUiStyles() {
  const data = require("@teambit/design.ui.styles.muted-italic");

  _designUiStyles = function () {
    return data;
  };

  return data;
}

function _designUiStyles2() {
  const data = require("@teambit/design.ui.styles.ellipsis");

  _designUiStyles2 = function () {
    return data;
  };

  return data;
}

function _scopeUiHooks() {
  const data = require("@teambit/scope.ui.hooks.use-scope");

  _scopeUiHooks = function () {
    return data;
  };

  return data;
}

function _componentsDrawerModule() {
  const data = _interopRequireDefault(require("./components-drawer.module.scss"));

  _componentsDrawerModule = function () {
    return data;
  };

  return data;
}

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

class ComponentsDrawer {
  constructor(treeNodeSlot) {
    this.treeNodeSlot = treeNodeSlot;
    (0, _defineProperty2().default)(this, "name", 'COMPONENTS');
    (0, _defineProperty2().default)(this, "render", () => {
      const {
        scope
      } = (0, _scopeUiHooks().useScopeQuery)();
      const {
        treeNodeSlot
      } = this;
      const TreeNodeRenderer = (0, _react().useCallback)(function TreeNode(props) {
        const children = props.node.children;
        if (!children) return /*#__PURE__*/_react().default.createElement(_uiFoundationUi().ComponentView, (0, _extends2().default)({}, props, {
          treeNodeSlot: treeNodeSlot
        })); // skip over scope node and render only children

        if (props.node.payload instanceof _uiFoundationUi().ScopePayload) {
          return /*#__PURE__*/_react().default.createElement(_react().default.Fragment, null, children.map(childNode => /*#__PURE__*/_react().default.createElement(TreeNodeRenderer, (0, _extends2().default)({
            key: childNode.id
          }, props, {
            node: childNode
          }))));
        }

        return /*#__PURE__*/_react().default.createElement(_uiFoundationUi().NamespaceTreeNode, props);
      }, [treeNodeSlot]);
      if (!scope) return /*#__PURE__*/_react().default.createElement(_uiFoundationUi2().FullLoader, null);
      if (scope.components.length === 0) return /*#__PURE__*/_react().default.createElement("span", {
        className: (0, _classnames().default)(_designUiStyles().mutedItalic, _designUiStyles2().ellipsis, _componentsDrawerModule().default.emptyScope)
      }, "Scope is empty");
      return /*#__PURE__*/_react().default.createElement(_uiFoundationUi().ComponentTree, {
        components: scope.components,
        TreeNode: TreeNodeRenderer
      });
    });
  }

}

exports.ComponentsDrawer = ComponentsDrawer;
