// @teambit/react/dist/templates/react-component-js/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.deprecatedReactComponentJS = exports.reactComponentJS = void 0;

function _component() {
  const data = require("./files/component");

  _component = function () {
    return data;
  };

  return data;
}

function _composition() {
  const data = require("./files/composition");

  _composition = function () {
    return data;
  };

  return data;
}

function _docs() {
  const data = require("./files/docs");

  _docs = function () {
    return data;
  };

  return data;
}

function _test() {
  const data = require("./files/test");

  _test = function () {
    return data;
  };

  return data;
}

const reactComponentJS = {
  name: 'react-js',
  description: 'a basic react component in js',
  generateFiles: context => {
    const {
      name,
      namePascalCase: Name
    } = context;
    const indexFile = {
      relativePath: 'index.js',
      content: `export { ${Name} } from './${name}';
`
    };
    return [indexFile, (0, _component().componentFile)(context), (0, _composition().compositionFile)(context), (0, _docs().docsFile)(context), (0, _test().testFile)(context)];
  }
};
exports.reactComponentJS = reactComponentJS;
const deprecatedReactComponentJS = {
  name: 'react-component-js',
  description: 'a basic react component in js',
  hidden: true,
  generateFiles: context => {
    const {
      name,
      namePascalCase: Name
    } = context;
    const indexFile = {
      relativePath: 'index.js',
      content: `export { ${Name} } from './${name}';
`
    };
    return [indexFile, (0, _component().componentFile)(context), (0, _composition().compositionFile)(context), (0, _docs().docsFile)(context), (0, _test().testFile)(context)];
  }
};
exports.deprecatedReactComponentJS = deprecatedReactComponentJS;
