// @teambit/linter/dist/linter.service.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

require("core-js/modules/es.promise.js");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.LinterService = void 0;

function _defineProperty2() {
  const data = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

  _defineProperty2 = function () {
    return data;
  };

  return data;
}

function _lodash() {
  const data = require("lodash");

  _lodash = function () {
    return data;
  };

  return data;
}

class LinterService {
  constructor(linterConfig) {
    this.linterConfig = linterConfig;
    (0, _defineProperty2().default)(this, "name", 'linter');
  }

  async run(context, options) {
    const mergedOpts = (0, _lodash().defaults)(options, this.linterConfig);
    const linterContext = Object.assign({}, {
      quiet: false,
      extensionFormats: mergedOpts.extensionFormats,
      fixTypes: mergedOpts.fixTypes,
      fix: mergedOpts.fix
    }, context);
    const linter = context.env.getLinter(linterContext);
    const results = await linter.lint(linterContext);
    return results;
  }

}

exports.LinterService = LinterService;
