// @yarnpkg/plugin-dlx/lib/commands/create.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const cli_1 = require("@yarnpkg/cli");
const core_1 = require("@yarnpkg/core");
const clipanion_1 = require("clipanion");
// eslint-disable-next-line arca/no-default-export
class CreateCommand extends cli_1.BaseCommand {
    constructor() {
        super(...arguments);
        this.pkg = clipanion_1.Option.String(`-p,--package`, {
            description: `The package to run the provided command from`,
        });
        this.quiet = clipanion_1.Option.Boolean(`-q,--quiet`, false, {
            description: `Only report critical errors instead of printing the full install logs`,
        });
        this.command = clipanion_1.Option.String();
        this.args = clipanion_1.Option.Proxy();
    }
    async execute() {
        const flags = [];
        if (this.pkg)
            flags.push(`--package`, this.pkg);
        if (this.quiet)
            flags.push(`--quiet`);
        const ident = core_1.structUtils.parseIdent(this.command);
        const modified = core_1.structUtils.makeIdent(ident.scope, `create-${ident.name}`);
        return this.cli.run([`dlx`, ...flags, core_1.structUtils.stringifyIdent(modified), ...this.args]);
    }
}
exports.default = CreateCommand;
CreateCommand.paths = [
    [`create`],
];
