// @teambit/react/dist/templates/react-app/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.reactAppTemplate = void 0;

function _index() {
  const data = require("./files/index");

  _index = function () {
    return data;
  };

  return data;
}

function _doc() {
  const data = require("./files/doc");

  _doc = function () {
    return data;
  };

  return data;
}

function _main() {
  const data = require("./files/main.runtime");

  _main = function () {
    return data;
  };

  return data;
}

function _compositions() {
  const data = require("./files/compositions");

  _compositions = function () {
    return data;
  };

  return data;
}

function _aspect() {
  const data = require("./files/aspect");

  _aspect = function () {
    return data;
  };

  return data;
}

function _app() {
  const data = require("./files/app");

  _app = function () {
    return data;
  };

  return data;
}

function _appRoot() {
  const data = require("./files/app-root");

  _appRoot = function () {
    return data;
  };

  return data;
}

const reactAppTemplate = {
  name: 'react-app',
  hidden: true,
  description: 'Creates a React app',
  generateFiles: context => {
    return [{
      relativePath: 'index.ts',
      content: (0, _index().indexFile)(context),
      isMain: true
    }, {
      relativePath: `${context.name}.docs.mdx`,
      content: (0, _doc().docFile)(context)
    }, {
      relativePath: `${context.name}.main.runtime.ts`,
      content: (0, _main().mainRuntimeFile)(context)
    }, {
      relativePath: `${context.name}.compositions.tsx`,
      content: (0, _compositions().compositionsFile)(context)
    }, {
      relativePath: `${context.name}.aspect.ts`,
      content: (0, _aspect().aspectFile)(context)
    }, {
      relativePath: `app.tsx`,
      content: (0, _app().appFile)(context)
    }, {
      relativePath: `${context.name}.app-root.tsx`,
      content: (0, _appRoot().appRootFile)(context)
    }];
  }
};
exports.reactAppTemplate = reactAppTemplate;
