// @teambit/legacy/dist/scope/graph/build-graph-from-fs.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

require("core-js/modules/es.array.iterator.js");

require("core-js/modules/es.promise.js");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.GraphFromFsBuilder = void 0;

function _defineProperty2() {
  const data = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

  _defineProperty2 = function () {
    return data;
  };

  return data;
}

function _pMapSeries() {
  const data = _interopRequireDefault(require("p-map-series"));

  _pMapSeries = function () {
    return data;
  };

  return data;
}

function _lodash() {
  const data = require("lodash");

  _lodash = function () {
    return data;
  };

  return data;
}

function _bitIds() {
  const data = _interopRequireDefault(require("../../bit-id/bit-ids"));

  _bitIds = function () {
    return data;
  };

  return data;
}

function _graph() {
  const data = _interopRequireDefault(require("../../scope/graph/graph"));

  _graph = function () {
    return data;
  };

  return data;
}

function _logger() {
  const data = _interopRequireDefault(require("../../logger/logger"));

  _logger = function () {
    return data;
  };

  return data;
}

function _exceptions() {
  const data = require("../exceptions");

  _exceptions = function () {
    return data;
  };

  return data;
}

function _generalError() {
  const data = _interopRequireDefault(require("../../error/general-error"));

  _generalError = function () {
    return data;
  };

  return data;
}

function _scopeComponentsImporter() {
  const data = _interopRequireDefault(require("../component-ops/scope-components-importer"));

  _scopeComponentsImporter = function () {
    return data;
  };

  return data;
}

class GraphFromFsBuilder {
  constructor(consumer, ignoreIds = new (_bitIds().default)(), loadComponentsFunc) {
    this.consumer = consumer;
    this.ignoreIds = ignoreIds;
    this.loadComponentsFunc = loadComponentsFunc;
    (0, _defineProperty2().default)(this, "graph", new (_graph().default)());
    (0, _defineProperty2().default)(this, "completed", []);
    (0, _defineProperty2().default)(this, "depth", 1);
    (0, _defineProperty2().default)(this, "shouldThrowOnMissingDep", true);
  }
  /**
   * create a graph with all dependencies and flattened dependencies of the given components.
   * the nodes are components and the edges has a label of the dependency type.
   *
   * the way how it is done is iterations by depths. each depth we gather all the dependencies of
   * that depths, make sure all objects exist and then check their dependencies for the next depth.
   * once there is no dependency left, we're on the last depth level and the graph is ready.
   *
   * for example, imagine the following graph:
   * A1 -> A2 -> A3
   * B1 -> B2 -> B3
   * C1 -> C2 -> C3
   *
   * where the buildGraph is given [A1, B1, C1].
   * first, it saves all these components as nodes in the graph. then, it finds the dependencies of
   * the next level, in this case they're [A2, B2, C2]. it runs `importMany` in case some objects
   * are missing. then, it loads them all (some from FS, some from the model) and sets the edges
   * between the component and the dependencies.
   * once done, it finds all their dependencies, which are [A3, B3, C3] and repeat the process
   * above. since there are no more dependencies, the graph is completed.
   * in this case, the total depth levels are 3.
   *
   * even with a huge project, there are not many depth levels. by iterating through depth levels
   * we keep performance sane as the importMany doesn't run multiple time and therefore the round
   * trips to the remotes are minimal.
   *
   * normally, one importMany of the seeders is enough as importMany knows to fetch all flattened.
   * however, since this buildGraph is performed on the workspace, a dependency may be new or
   * modified and as such, we don't know its flattened yet.
   */


  async buildGraph(ids) {
    _logger().default.debug(`GraphFromFsBuilder, buildGraph with ${ids.length} seeders`);

    const components = await this.loadManyComponents(ids, '<none>');
    await this.processManyComponents(components);
    return this.graph;
  }

  getAllDeps(component) {
    return component.getAllDependenciesIds().difference(this.ignoreIds);
  }

  async processManyComponents(components) {
    _logger().default.debug(`GraphFromFsBuilder.processManyComponents depth ${this.depth}, ${components.length} components`);

    this.depth += 1;
    const allDeps = (0, _lodash().flatten)(components.map(c => this.getAllDeps(c)));
    const allDepsWithScope = allDeps.filter(dep => dep.hasScope());
    const scopeComponentsImporter = new (_scopeComponentsImporter().default)(this.consumer.scope);
    await scopeComponentsImporter.importMany(_bitIds().default.uniqFromArray(allDepsWithScope), undefined, this.shouldThrowOnMissingDep);
    const allDependencies = await (0, _pMapSeries().default)(components, component => this.processOneComponent(component));
    const allDependenciesFlattened = (0, _lodash().flatten)(allDependencies);
    if (allDependenciesFlattened.length) await this.processManyComponents(allDependenciesFlattened);
  }

  async processOneComponent(component) {
    const idStr = component.id.toString();
    if (this.completed.includes(idStr)) return [];
    const allIds = this.getAllDeps(component);
    const allDependencies = await this.loadManyComponents(allIds, idStr);
    Object.entries(component.depsIdsGroupedByType).forEach(([depType, depsIds]) => {
      depsIds.forEach(depId => {
        if (this.ignoreIds.has(depId)) return;

        if (!this.graph.hasNode(depId.toString())) {
          throw new Error(`buildOneComponent: missing node of ${depId.toString()}`);
        }

        this.graph.setEdge(idStr, depId.toString(), depType);
      });
    });
    this.completed.push(idStr);
    return allDependencies;
  }

  async loadManyComponents(componentsIds, dependenciesOf) {
    return (0, _pMapSeries().default)(componentsIds, async comp => {
      const idStr = comp.toString();
      const fromGraph = this.graph.node(idStr);
      if (fromGraph) return fromGraph;

      try {
        const component = await this.loadComponent(comp);
        this.graph.setNode(idStr, component);
        return component;
      } catch (err) {
        if (err instanceof _exceptions().ComponentNotFound) {
          throw new (_generalError().default)(`error: component "${idStr}" was not found.\nthis component is a dependency of "${dependenciesOf}" and is needed as part of the graph generation`);
        }

        _logger().default.error(`failed loading dependencies of ${dependenciesOf}`);

        throw err;
      }
    });
  }

  async loadComponent(componentId) {
    const componentMap = this.consumer.bitMap.getComponentIfExist(componentId);
    const isOnWorkspace = Boolean(componentMap);

    if (isOnWorkspace) {
      if (this.loadComponentsFunc) {
        const dependency = await this.loadComponentsFunc([componentId]);

        if (!dependency.length || !dependency[0]) {
          throw new Error(`unable to load ${componentId.toString()} using custom load function`);
        }

        return dependency[0];
      }

      return this.consumer.loadComponentForCapsule(componentId);
    } // a dependency might have been installed as a package in the workspace, and as such doesn't
    // have a componentMap.


    const componentFromModel = await this.consumer.loadComponentFromModel(componentId);
    return componentFromModel.clone();
  }

}

exports.GraphFromFsBuilder = GraphFromFsBuilder;
