// @teambit/changelog/dist/ui/change-log-page.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ChangeLogPage = ChangeLogPage;

function _component() {
  const data = require("@teambit/component");

  _component = function () {
    return data;
  };

  return data;
}

function _documenterUi() {
  const data = require("@teambit/documenter.ui.heading");

  _documenterUi = function () {
    return data;
  };

  return data;
}

function _designUi() {
  const data = require("@teambit/design.ui.separator");

  _designUi = function () {
    return data;
  };

  return data;
}

function _componentUi() {
  const data = require("@teambit/component.ui.version-block");

  _componentUi = function () {
    return data;
  };

  return data;
}

function _classnames() {
  const data = _interopRequireDefault(require("classnames"));

  _classnames = function () {
    return data;
  };

  return data;
}

function _componentUiHooks() {
  const data = require("@teambit/component.ui.hooks.use-snaps");

  _componentUiHooks = function () {
    return data;
  };

  return data;
}

function _mdxUi() {
  const data = require("@teambit/mdx.ui.mdx-layout");

  _mdxUi = function () {
    return data;
  };

  return data;
}

function _componentInstructions() {
  const data = require("@teambit/component.instructions.exporting-components");

  _componentInstructions = function () {
    return data;
  };

  return data;
}

function _designUi2() {
  const data = require("@teambit/design.ui.alert-card");

  _designUi2 = function () {
    return data;
  };

  return data;
}

function _react() {
  const data = _interopRequireWildcard(require("react"));

  _react = function () {
    return data;
  };

  return data;
}

function _changeLogPageModule() {
  const data = _interopRequireDefault(require("./change-log-page.module.scss"));

  _changeLogPageModule = function () {
    return data;
  };

  return data;
}

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function ChangeLogPage({
  className
}) {
  var _snaps$, _snaps$2;

  const component = (0, _react().useContext)(_component().ComponentContext);
  const {
    snaps,
    loading
  } = (0, _componentUiHooks().useSnaps)(component.id);
  if (!snaps) return null;

  if (snaps.length === 0 && !loading) {
    return /*#__PURE__*/_react().default.createElement("div", {
      className: (0, _classnames().default)(_changeLogPageModule().default.changeLogPage, className)
    }, /*#__PURE__*/_react().default.createElement(_documenterUi().H1, {
      className: _changeLogPageModule().default.title
    }, "History"), /*#__PURE__*/_react().default.createElement(_designUi().Separator, {
      isPresentational: true,
      className: _changeLogPageModule().default.separatorNoChangeLog
    }), /*#__PURE__*/_react().default.createElement(_designUi2().AlertCard, {
      level: "info",
      title: "There is no change log as this component has not been exported yet. Learn how to export components:",
      className: _changeLogPageModule().default.changeLogCard
    }, /*#__PURE__*/_react().default.createElement(_mdxUi().MDXLayout, null, /*#__PURE__*/_react().default.createElement(_componentInstructions().ExportingComponents, null))));
  }

  const latestVersion = ((_snaps$ = snaps[0]) === null || _snaps$ === void 0 ? void 0 : _snaps$.tag) || ((_snaps$2 = snaps[0]) === null || _snaps$2 === void 0 ? void 0 : _snaps$2.hash);
  return /*#__PURE__*/_react().default.createElement("div", {
    className: (0, _classnames().default)(_changeLogPageModule().default.changeLogPage, className)
  }, /*#__PURE__*/_react().default.createElement(_documenterUi().H1, {
    className: _changeLogPageModule().default.title
  }, "History"), /*#__PURE__*/_react().default.createElement(_designUi().Separator, {
    isPresentational: true,
    className: _changeLogPageModule().default.separator
  }), snaps.map((snap, index) => {
    const isLatest = latestVersion === snap.tag || latestVersion === snap.hash;
    return /*#__PURE__*/_react().default.createElement(_componentUi().VersionBlock, {
      key: index,
      componentId: component.id.fullName,
      isLatest: isLatest,
      snap: snap
    });
  }));
}
