// @teambit/schema/dist/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Parser", {
  enumerable: true,
  get: function () {
    return _parser().Parser;
  }
});
Object.defineProperty(exports, "SchemaExtractor", {
  enumerable: true,
  get: function () {
    return _schemaExtractor().SchemaExtractor;
  }
});
Object.defineProperty(exports, "Export", {
  enumerable: true,
  get: function () {
    return _schemas().Export;
  }
});
Object.defineProperty(exports, "Module", {
  enumerable: true,
  get: function () {
    return _schemas().Module;
  }
});
Object.defineProperty(exports, "SchemaAspect", {
  enumerable: true,
  get: function () {
    return _schema().SchemaAspect;
  }
});
Object.defineProperty(exports, "SemanticSchema", {
  enumerable: true,
  get: function () {
    return _semanticSchema().SemanticSchema;
  }
});

function _parser() {
  const data = require("./parser");

  _parser = function () {
    return data;
  };

  return data;
}

function _schemaExtractor() {
  const data = require("./schema-extractor");

  _schemaExtractor = function () {
    return data;
  };

  return data;
}

function _schemas() {
  const data = require("./schemas");

  _schemas = function () {
    return data;
  };

  return data;
}

function _schema() {
  const data = require("./schema.aspect");

  _schema = function () {
    return data;
  };

  return data;
}

function _semanticSchema() {
  const data = require("./semantic-schema");

  _semanticSchema = function () {
    return data;
  };

  return data;
}
